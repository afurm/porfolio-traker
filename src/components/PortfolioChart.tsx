import React, { useState } from 'react';
import {
  PieChart,
  Pie,
  Tooltip,
  Cell,
  Legend,
  BarChart,
  Bar,
  XAxis,
  YAxis,
  CartesianGrid,
} from 'recharts';
import { Button, Box, Container } from '@mui/material';

export interface PortfolioChartProps {
  assets: { coinName: string; value: number }[];
}

const COLORS = ['#0088FE', '#00C49F', '#FFBB28', '#FF8042', '#8884d8'];

const PortfolioChart: React.FC<PortfolioChartProps> = ({ assets }) => {
  const [chartType, setChartType] = useState<'pie' | 'bar'>('pie');

  const handleChartToggle = () => {
    setChartType((prev) => (prev === 'pie' ? 'bar' : 'pie'));
  };

  return (
    <Container>
      <Box display="flex" justifyContent="center" marginBottom={2}>
        <Button variant="contained" color="primary" onClick={handleChartToggle}>
          Toggle to {chartType === 'pie' ? 'Bar' : 'Pie'} Chart
        </Button>
      </Box>
      {chartType === 'pie' ? (
        <PieChart width={400} height={400}>
          <Pie
            data={assets}
            cx={200}
            cy={200}
            label={({ name, value }) => `${name}: $${value}`}
            outerRadius={150}
            fill="#8884d8"
            dataKey="value"
            nameKey="coinName"
          >
            {assets.map((_, index) => (
              <Cell key={`cell-${index}`} fill={COLORS[index % COLORS.length]} />
            ))}
          </Pie>
          <Tooltip formatter={(value: number, name: string) => [`$${value}`, name]} />
          <Legend />
        </PieChart>
      ) : (
        <BarChart
          width={500}
          height={300}
          data={assets}
          margin={{ top: 20, right: 30, left: 20, bottom: 5 }}
        >
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis dataKey="coinName" />
          <YAxis />
          <Tooltip formatter={(value: number, name: string) => [`$${value}`, name]} />
          <Legend />
          <Bar dataKey="value" fill="#8884d8">
            {assets.map((_, index) => (
              <Cell key={`cell-${index}`} fill={COLORS[index % COLORS.length]} />
            ))}
          </Bar>
        </BarChart>
      )}
    </Container>
  );
};

export default PortfolioChart;
