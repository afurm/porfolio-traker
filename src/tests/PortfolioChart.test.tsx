import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import PortfolioChart, { PortfolioChartProps } from '@/components/PortfolioChart';

describe('PortfolioChart Component Snapshot Test', () => {
  it('renders correctly with Pie chart initially', () => {
    const mockData: PortfolioChartProps['assets'] = [
      { coinName: 'Bitcoin', value: 50 },
      { coinName: 'Ethereum', value: 30 },
      { coinName: 'Litecoin', value: 20 },
    ];

    const { asFragment, getByText } = render(<PortfolioChart assets={mockData} />);

    // Ensure the button text is "Toggle to Bar Chart"
    expect(getByText('Toggle to Bar Chart')).toBeInTheDocument();

    // Take a snapshot of the initial rendering (Pie chart)
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when toggled to Bar chart', () => {
    const mockData: PortfolioChartProps['assets'] = [
      { coinName: 'Bitcoin', value: 50 },
      { coinName: 'Ethereum', value: 30 },
      { coinName: 'Litecoin', value: 20 },
    ];

    const { asFragment, getByText } = render(<PortfolioChart assets={mockData} />);

    // Simulate button click to toggle to Bar chart
    const toggleButton = getByText('Toggle to Bar Chart');
    fireEvent.click(toggleButton);

    // Ensure the button text updates to "Toggle to Pie Chart"
    expect(getByText('Toggle to Pie Chart')).toBeInTheDocument();

    // Take a snapshot after toggling (Bar chart)
    expect(asFragment()).toMatchSnapshot();
  });
});
