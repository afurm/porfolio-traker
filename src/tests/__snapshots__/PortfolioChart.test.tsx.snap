// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioChart Component Snapshot Test renders correctly when toggled to Bar chart 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1og1qar"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-142372x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Toggle to Pie Chart
      </button>
    </div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 500px; height: 300px;"
    >
      <svg
        class="recharts-surface"
        height="300"
        style="width: 100%; height: 100%;"
        viewBox="0 0 500 300"
        width="500"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts5-clip"
          >
            <rect
              height="245"
              width="390"
              x="80"
              y="20"
            />
          </clippath>
        </defs>
        <g
          class="recharts-cartesian-grid"
        >
          <g
            class="recharts-cartesian-grid-horizontal"
          >
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="80"
              x2="470"
              y="20"
              y1="265"
              y2="265"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="80"
              x2="470"
              y="20"
              y1="203.75"
              y2="203.75"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="80"
              x2="470"
              y="20"
              y1="142.5"
              y2="142.5"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="80"
              x2="470"
              y="20"
              y1="81.25"
              y2="81.25"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="80"
              x2="470"
              y="20"
              y1="20"
              y2="20"
            />
          </g>
          <g
            class="recharts-cartesian-grid-vertical"
          >
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="145"
              x2="145"
              y="20"
              y1="20"
              y2="265"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="275"
              x2="275"
              y="20"
              y1="20"
              y2="265"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="405"
              x2="405"
              y="20"
              y1="20"
              y2="265"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="80"
              x2="80"
              y="20"
              y1="20"
              y2="265"
            />
            <line
              fill="none"
              height="245"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="390"
              x="80"
              x1="470"
              x2="470"
              y="20"
              y1="20"
              y2="265"
            />
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            orientation="bottom"
            stroke="#666"
            width="390"
            x="80"
            x1="80"
            x2="470"
            y="265"
            y1="265"
            y2="265"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="390"
                x="80"
                x1="145"
                x2="145"
                y="265"
                y1="271"
                y2="265"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="390"
                x="145"
                y="273"
              >
                <tspan
                  dy="0.71em"
                  x="145"
                >
                  Bitcoin
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="390"
                x="80"
                x1="275"
                x2="275"
                y="265"
                y1="271"
                y2="265"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="390"
                x="275"
                y="273"
              >
                <tspan
                  dy="0.71em"
                  x="275"
                >
                  Ethereum
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="390"
                x="80"
                x1="405"
                x2="405"
                y="265"
                y1="271"
                y2="265"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="390"
                x="405"
                y="273"
              >
                <tspan
                  dy="0.71em"
                  x="405"
                >
                  Litecoin
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="245"
            orientation="left"
            stroke="#666"
            width="60"
            x="20"
            x1="80"
            x2="80"
            y="20"
            y1="20"
            y2="265"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="245"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="265"
                y2="265"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="245"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="265"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="245"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="203.75"
                y2="203.75"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="245"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="203.75"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  15
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="245"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="142.5"
                y2="142.5"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="245"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="142.5"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  30
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="245"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="81.25"
                y2="81.25"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="245"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="81.25"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  45
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="245"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="20"
                y2="20"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="245"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="20"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  60
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-bar"
        >
          <g
            class="recharts-layer recharts-bar-rectangles"
          >
            <g
              class="recharts-layer"
            >
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-legend-wrapper"
        style="position: absolute; width: 450px; height: auto; left: 20px; bottom: 5px;"
      >
        <ul
          class="recharts-default-legend"
          style="padding: 0px; margin: 0px; text-align: center;"
        >
          <li
            class="recharts-legend-item legend-item-0"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              viewBox="0 0 32 32"
              width="14"
            >
              <title />
              <desc />
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#8884d8"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
              style="color: rgb(136, 132, 216);"
            >
              value
            </span>
          </li>
        </ul>
      </div>
      <div
        class="recharts-tooltip-wrapper"
        style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PortfolioChart Component Snapshot Test renders correctly with Pie chart initially 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1og1qar"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-142372x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Toggle to Bar Chart
      </button>
    </div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 400px; height: 400px;"
    >
      <svg
        class="recharts-surface"
        cx="50%"
        cy="50%"
        height="400"
        style="width: 100%; height: 100%;"
        viewBox="0 0 400 400"
        width="400"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="390"
              width="390"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
          tabindex="0"
        >
          <g
            class="recharts-layer"
          />
        </g>
      </svg>
      <div
        class="recharts-legend-wrapper"
        style="position: absolute; width: 390px; height: auto; left: 5px; bottom: 5px;"
      >
        <ul
          class="recharts-default-legend"
          style="padding: 0px; margin: 0px; text-align: center;"
        >
          <li
            class="recharts-legend-item legend-item-0"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              viewBox="0 0 32 32"
              width="14"
            >
              <title />
              <desc />
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#0088FE"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
              style="color: rgb(0, 136, 254);"
            >
              Bitcoin
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-1"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              viewBox="0 0 32 32"
              width="14"
            >
              <title />
              <desc />
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#00C49F"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
              style="color: rgb(0, 196, 159);"
            >
              Ethereum
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-2"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              viewBox="0 0 32 32"
              width="14"
            >
              <title />
              <desc />
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#FFBB28"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
              style="color: rgb(255, 187, 40);"
            >
              Litecoin
            </span>
          </li>
        </ul>
      </div>
      <div
        class="recharts-tooltip-wrapper"
        style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
